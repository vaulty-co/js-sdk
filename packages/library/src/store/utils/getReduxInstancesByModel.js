import invariant from 'invariant';
import { createAction, handleActions } from 'redux-actions';
import produce from 'immer';

/**
 * Inner invariant with specified prefix
 * @param {boolean} condition
 * @param {string} message
 */
const innerInvariant = (condition, message) => invariant(
  condition,
  `[getReduxInstances] ${message}`,
);

/**
 * Convert methods list to actions
 * @param {Function} Model - constructor of model
 * @param {Array<string>} methods
 * @return {Object}
 */
const methodsToActions = (Model, methods) => (
  methods.reduce((resultActions, method) => {
    innerInvariant(
      Model.prototype[method],
      `Can not create action but unspecified method '${method}' for model '${Model.name}'`,
    );

    return {
      [method]: createAction(method),
      ...resultActions,
    };
  }, {})
);

/**
 * Get reducer and actions by model instance and methods list
 * @param {Function} Model - constructor of model
 * @param {Array<string>} methods - list of methods, which should be used by actions
 * @param {Object} initialState
 * @return {{reducer: Function, actions: Object}}
 */
const getReduxInstancesByModel = (Model, methods = [], initialState) => {
  const actions = methodsToActions(Model, methods);
  const reducer = handleActions(
    methods.reduce((resultReducer, method) => ({
      [method]: produce(
        (state, { payload }) => state[method](payload),
      ),
      ...resultReducer,
    }), {}),
    initialState,
  );
  return {
    actions,
    reducer,
  };
};

export default {
  getReduxInstancesByModel,
};
export {
  getReduxInstancesByModel,
};
