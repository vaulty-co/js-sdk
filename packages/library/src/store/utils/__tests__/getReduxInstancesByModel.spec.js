import { getReduxInstancesByModel } from '../getReduxInstancesByModel';

describe('getReduxInstancesByModel', () => {
  class Model {
    constructor() {
      this.items = {};
    }

    addItem(item) {
      this.items[item.id] = item;
    }

    removeItem(item) {
      delete this.items[item.id];
    }
  }

  it('should create actions by models methods', () => {
    // Arrange
    const { actions } = getReduxInstancesByModel(Model, ['addItem'], new Model());

    // Act
    const addItemAction = actions.addItem({ id: 'item-id' });

    // Assert
    expect(addItemAction).toMatchObject({
      type: 'addItem',
      payload: { id: 'item-id' },
    });
  });

  it('should not create actions for methods, which are not specified in list', () => {
    // Arrange
    const { actions } = getReduxInstancesByModel(Model, ['addItem'], new Model());

    // Assert
    expect(actions).not.toHaveProperty('removeItem');
  });

  it('should throw error, when methods do not exist in model', () => {
    // Act
    const throwError = () => (
      getReduxInstancesByModel(Model, ['someNonExistMethod'], new Model())
    );

    // Assert
    expect(throwError).toThrowErrorMatchingSnapshot();
  });

  it('should create reducer by models method, which call model method by action', () => {
    // Arrange
    const { actions, reducer } = getReduxInstancesByModel(Model, ['addItem'], new Model());
    const item = { id: 'item-id' };

    // Act
    const expectedResult = reducer(new Model(), actions.addItem(item));

    // Assert
    expect(expectedResult.items['item-id']).toBe(item);
  });

  it('should use provided model instance as initial state', () => {
    // Arrange
    const initialState = new Model();
    const { reducer } = getReduxInstancesByModel(Model, ['addItem'], initialState);

    // Act
    const expectedResult = reducer(undefined, { type: 'unknownAction' });

    // Assert
    expect(expectedResult).toBe(initialState);
  });
});
