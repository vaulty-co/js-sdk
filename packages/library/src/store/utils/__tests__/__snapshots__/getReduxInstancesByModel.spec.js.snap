// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getReduxInstancesByModel should throw error, when methods do not exist in model 1`] = `"[getReduxInstances] Can not create action for unspecified method 'someNonExistMethod' of model 'Model'"`;
